sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala

```diff
@@ -0,0 +1,164 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import javax.annotation.Nullable
+
+import scala.annotation.tailrec
+
+import org.apache.spark.sql.catalyst.expressions.{BinaryExpression, Cast, Collate, ComplexTypeMergingExpression, CreateArray, ExpectsInputTypes, Expression, Predicate, SortOrder}
+import org.apache.spark.sql.errors.QueryCompilationErrors
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{AbstractDataType, ArrayType, DataType, StringType}
+
+object CollationTypeCasts extends TypeCoercionRule {
+  override val transform: PartialFunction[Expression, Expression] = {
+    case e if !e.childrenResolved => e
+    case sc @ (_: BinaryExpression
```

Comment:
@cloud-fan Is this what you meant? Now this type of query will be broken `SELECT array('b' collate unicode_ci) || 'a' collate unicode`, but I can add support for array explicit as a followup.

Reply:
Also, I do not believe incorporating string collations into core functions is the best way to go. If you take a look at findWiderCommonForTwo, it is implemented differently for AnsiTypeCoercion and TypeCoercion. Firstly both of those take leftFold and go through expressions in order, but in our case that is not what we want. We want to first know what collations are explicit and then look for anything else. This is important because we want to have clear design on which error has to come first, and explicit mismatch should always come before implicit one. Another thing is that AnsiTypeCoercion seems to behave differently for StringTypes, we do not reorder any type in the sequence, which would result in implementing the same thing we did here with addition of another rule, but just in a core, already well defined function.
One more thing I would add is that not all rules actually use core functions, e.g. ConcatCoercion. Also doing collations casting in one rule is way more efficient, as otherwise we would have to reorder StringTypes in every other rule, to make sure explicit collation mismatches are thrown first, which would result in multiple reorderings as opposed to constant of 2, one at the beginning for all StringTypes and the other at the end for all expressions that were cast to StringType from other types in some other implicit casting rules.

Reply:
This is a good point that we can't reuse the existing `fold` style to do implicit cast for string collations. If we must rewrite the code, a new rule sounds better.

To avoid the notorious analyzer rule execution order issues, let's make sure this new rule is orthogonal to existing implicit cast rules: the existing "fold" style implicit cast should not apply to string type with different collations.

Reply:
I went through all the core functions and it seems like StringType implicit castings are already blocked, or go to default branch which is None. Is there something else we want to check?